### PR TITLE
🔨(project) warn user when running make bootstrap from root user

### PR DESCRIPTION
**Pull request by slandrault**
## Purpose

In order to not mess up user filesystem, the build should not be run as a root user.
This commit add a `check-root-user` make target that will check if the user running the make is not with an id = 0 (root).

## Proposal

To guard any make target from being run as root, a `check-root-user` target can be called first. If the user is root, then the command will fail with a warning message. The warning/fail behavior can be override by setting a `ALLOW_ROOT` variable like: 
```
$ sudo ALLOW_ROOT=1 make bootstrap
```

Fixes https://github.com/openfun/openedx-docker/issues/204

### CODE REVIEW

**Comment by slandrault** [comment]
For the moment, the solution is implemented as an "fail if run as root". I'll evaluate the feasibility of achieving similar feature with a warning+confirmation instead of failing all the time.

**Comment by slandrault** [comment]
The target has been updated. In case of root call, the command will now ask for confirmation before continuation instead of leaving.

File: releases/dogwood/3/fun/Dockerfile [hunk 1]
```diff
@@ -215,12 +215,8 @@ RUN apt-get update && \
...
 RUN groupadd --gid ${DOCKER_GID} edx || \
       echo "Group with ID ${DOCKER_GID} already exists." && \
-    useradd \
-      --create-home \
-      --home-dir /home/edx \
-      --uid ${DOCKER_UID} \
-      --gid ${DOCKER_GID} \
-      edx && \
+    useradd --create-home --home-dir /home/edx --uid ${DOCKER_UID} --gid ${DOCKER_GID} edx || \
+      echo "User with ID ${DOCKER_UID} already exists." && \
```
Comment (jmaupetit) [inline releases/dogwood/3/fun/Dockerfile:219]: We are not sure of the reason why the useradd command fails:

```suggestion
      echo "Skip user creation (user with ID ${DOCKER_UID} already exists?)" && \
```

Comment (slandrault) [reply releases/dogwood/3/fun/Dockerfile:219]: Good point

**Review by jmaupetit — COMMENTED** [review]
Maybe other commands that create/access mounted volumes should also have the `ensure-non-root` dependency (_e.g._ `build`, `build-dev`, `tree`), WDYT?

**Comment by slandrault** [comment]
> Maybe other commands that create/access mounted volumes should also have the `ensure-non-root` dependency (_e.g._ `build`, `build-dev`, `tree`), WDYT?

They could, but that means that when bootstrap will be run it will ask for confirmation when reaching each of those target... Maybe not very user friendly?

**Comment by jmaupetit** [comment]
> Maybe not very user friendly?

At least the user will understand that this is not the recommended way to do things :joy:

**Comment by slandrault** [comment]
A side effect would be that the bootstrap won't be able to run by itself until completion because the user would have to validate several times that he is ok to continue as sudo...

**Comment by sampaccoud** [comment]
We could request to set an env variable to be allowed to run any command as sudo?

**Comment by slandrault** [comment]
> We could request to set an env variable to be allowed to run any command as sudo?

:+1:  Sounds like a an elegant solution. If the variable is set then go, if not then warning message and fail nicely.

File: Makefile [hunk 2]
```diff
@@ -342,6 +357,19 @@ tree: \
 tree:  ## create data directories mounted as volumes
 .PHONY: tree
 
+verify-root:  # Make sure the user calling this is not currently root
```
Comment (lunika) [inline Makefile:360]: In your commit message you called it `ensure-not-root`

Comment (lunika) [inline Makefile:360]:
```suggestion
verify-root:  ## Make sure the user calling this is not currently root
```

File: Makefile [hunk 3]
```diff
@@ -166,14 +168,19 @@ create-symlinks:  ## create symlinks to local configuration (mounted via a volum
 	  ln -sf /config/lms/root_urls.py /edx/app/edxapp/edx-platform/lms/"
 .PHONY: create-symlinks
 
-demo-course: fetch-demo  ## import demo course from edX repository
+# Import demo course from edX repository
```
Comment (lunika) [inline Makefile:171]:
```suggestion
```

File: Makefile [hunk 4]
```diff
@@ -166,14 +168,19 @@ create-symlinks:  ## create symlinks to local configuration (mounted via a volum
...
+demo-course: \
+	verify-root \
+	fetch-demo
+demo-course:
```
Comment (lunika) [inline Makefile:175]:
```suggestion
demo-course: ## import demo course from edX repository
```

File: Makefile [hunk 5]
```diff
@@ -231,7 +240,10 @@ dev-ui-toolkit:
 	  bash -c "cd node_modules/edx-ui-toolkit || exit 0 && npm install"
 .PHONY: dev-ui-toolkit
 
-dev-watch: tree  ## start assets watcher (front-end development)
+# Start assets watcher (front-end development)
```
Comment (lunika) [inline Makefile:243]:
```suggestion

```

File: Makefile [hunk 6]
```diff
@@ -231,7 +240,10 @@ dev-ui-toolkit:
...
+dev-watch: \
+	verify-root \
+	tree \
```
Comment (lunika) [inline Makefile:246]:
```suggestion
	tree
dev-watch: ## Start assets watcher (front-end development)
```

File: releases/dogwood/3/bare/Dockerfile [hunk 7]
```diff
@@ -223,12 +223,8 @@ RUN apt-get update && \
 
 RUN groupadd --gid ${DOCKER_GID} edx || \
       echo "Group with ID ${DOCKER_GID} already exists." && \
-    useradd \
-      --create-home \
-      --home-dir /home/edx \
-      --uid ${DOCKER_UID} \
-      --gid ${DOCKER_GID} \
-      edx && \
+    useradd --create-home --home-dir /home/edx --uid ${DOCKER_UID} --gid ${DOCKER_GID} edx || \
```
Comment (lunika) [inline releases/dogwood/3/bare/Dockerfile:226]: You should keep the multiline command, we use this format to improve readability

File: .circleci/config.yml [hunk 8]
```diff
@@ -144,17 +144,33 @@ jobs:
   #
   # Note that the job name should match the EDX_RELEASE value
```
Comment (lunika) [inline .circleci/config.yml:145]: in your commit title: 

`s/prevent make bootstrap to be run from root user/prevent make bootstrap to be run as root user/`

**Review by lunika — COMMENTED** [review]
The makefile is self documented with the `help` task. All comments starting by `##` after a task name must be kept and you will see them when you run `make` or `make help`

File: Makefile (hunk 2 above)
Comment (jmaupetit) [reply Makefile:360]: What about `check-root-user`?

File: Makefile (hunk 3 above)
Comment (slandrault) [reply Makefile:171]: If we do that then we break the "codestyle" of the file. Everytime a target is on multiple line then the comment is above the command, only when there is no / the comment is next to the target name.

Comment (lunika) [reply Makefile:183]: Take a look at the `help` task, this task is using comments starting with `##`

Comment (lunika) [reply Makefile:183]: Also look this article https://marmelab.com/blog/2016/02/29/auto-documented-makefile.html

Comment (slandrault) [reply Makefile]: Ohhhh ok yes I understand why now! Thx!

File: Makefile [hunk 9]
```diff
@@ -145,6 +147,18 @@ build:  ## build the edxapp production image
 	$(COMPOSE) build nginx
 .PHONY: build
 
+check-root-user:  ## Make sure the user calling this is not currently root
+	@if [ $(shell id -u) -eq 0 ]; \
```
Comment (jmaupetit) [inline Makefile:151]: It is generally safer to use the double bracket syntax:

```suggestion
	@if [[ $(shell id -u) -eq 0 ]]; \
```

Reference: http://mywiki.wooledge.org/BashGuide/Practices#Bash_Tests

**Review by jmaupetit — CHANGES_REQUESTED** [review]
Rules that may create files and thus should have the `check-root-user` dependency are:

- `build`
- `clean`
- `create-symlinks`
- `demo-course`
- `dev-assets`
- `dev-build`
- `dev-install`
- `dev-watch`
- `fetch-demo`
- `fetch-release`
- `migrate`
- `run`
- `tree`

File: Makefile [hunk 10]
```diff
@@ -114,6 +114,7 @@ $(FLAVORED_EDX_RELEASE_PATH)/src/edx-demo-course/README.md:
 	tar xzf /tmp/edx-demo.tgz -C $(FLAVORED_EDX_RELEASE_PATH)/src/edx-demo-course --strip-components=1
 
 bootstrap: \
+  check-root-user \
```
[the remark this answers is not in the document]
Comment (slandrault) [reply Makefile:117]: Fair point.

File: releases/dogwood/3/bare/Dockerfile [hunk 11]
```diff
@@ -222,13 +222,14 @@ RUN apt-get update && \
     rm -rf /var/lib/apt/lists/*
```
Comment (jmaupetit) [inline releases/dogwood/3/bare/Dockerfile:222]: Your commit message scope should be: `flavors/*` and I would rephrase your commit title as:

:hammer:(flavors/*) allow builds using the root user

If you hesitate on a scope, check the project's commits history!

**Review by jmaupetit — APPROVED** [review]
